doc/mqldump.1
.\" Man page for mqldump
.\" Use the following command to view man page:
.\"
.\"  tbl mqldump.1 | nroff -man | less
.\"
.TH MQLDUMP 1 "October 9, 2007"
.SH NAME
mqldump \- A tool to dump Emdros databases in MQL
.SH SYNOPSIS
\fBmqldump\fR [ options ] database [object type names...]
.br
.SH DESCRIPTION
\fBmqldump\fR is a tool to dump (parts of) an Emdros database into an
MQL script.  This MQL script can then later be run through the
\fBmql(1)\fR program to recreate (the dumped parts of) the database.

If one or more object type names are given after the database name,
only those object types are processed in the order given.  Otherwise,
all object types are processed.

If the -o option is given, the output is placed on the file with the
filename given.  Otherwise, the output is dumped to stdout.

Object id_ds are preserved, i.e., when dumping object data, each
CREATE OBJECT statement has a WITH ID_D part.  This makes it possible
to import the MQL script again and preserve things like id_d feature
values that point to other objects.

If the -v or --verbose option is given, a lot of information is
displayed showing the progress of the dump.  When dumping objects,
each emitted dot represents 5000 objects.

.SH OPTIONS
\fBmqldump\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help
.TP
.BI \-V\ ,\ \-\-version
show version
.TP
.BI \-b\ ,\ \-\-backend \ backend
set database backend to `backend'. Valid values are: For PostgreSQL:
"p", "pg", "postgres", and "postgresql". For MySQL: "m", "my", and
"mysql". For SQLite 2.X.X: "2", "s", "l", "lt", "sqlite", and
"sqlite2". For SQLite 3.X.X: "3", "s3", "lt3", and "sqlite3".
.TP
.BI \-h\ ,\ \-\-host \ hostname 
set db back-end hostname to connect to (default is 'localhost').  Has
no effect on SQLite.
.TP
.BI \-u\ ,\ \-\-user \ user
set database user to connect as (default is 'emdf').  Has no effect on
SQLite.
.TP
.BI \-p\ ,\ \-\-password \ password
set password to use for the database user.  Has no effect on SQLite,
unless you have an encryption-enabled SQLite, in which case this gets
passed as the key.

.TP
.BI \-v\ ,\ \-\-verbose
be verbose
.TP
.BI \-o \ filename
produce output on file rather than stdout
.TP
.BI \-n\ ,\ \-\-no\-create\-db
do not produce CREATE DATABASE and USE DATABASE statements at the beginning
.TP
.BI \-i\ ,\ \-\-no\-create\-info
do not produce meta-data info (same as --no-create-enums
--no-create-object-types)
.TP
.BI \-t\ ,\ \-\-no\-create\-object\-types
do not produce object type meta-data info (i.e., don't produce CREATE
OBJECT TYPE statements)
.TP
.BI \-e\ ,\ \-\-no\-create\-enums
do not produce enumeration (meta-)data info (i.e., don't create
enumerations and enumeration constants by producing CREATE ENUMERATION
statements)
.TP
.BI \-m\ ,\ \-\-no\-create\-monad\-sets
do not produce arbitrary monad sets (i.e., don't create monad sets by 
producing CREATE MONAD SET statements)
.TP
.BI \-d\ ,\ \-\-no\-create\-objects
do not produce object data (i.e., don't produce CREATE OBJECT statements)
.TP
.BI \-\-batch\-create\-objects
use CREATE OBJECTS WITH OBJECT TYPE rather than individual CREATE
OBJECT statements.
.TP
.BI \-\-batch\-size \ number 
issue \fInumber\fR object-creations per CREATE OBJECTS WITH OBJECT
TYPE statement.  Implies \fB\-\-batch\-create\-objects\fR.
.TP
.BI \-\-start \ number 
set monad at which dumping should start.  Default is MIN_M (i.e.,
lowest monad in database).
.TP
.BI \-\-end \ number 
set monad at which dumping should end.  Default is MAX_M (i.e.,
highest monad in database).
.TP
.BI \-\-encoding \ encoding
When issuing the CREATE DATABASE statement, pass the 'encoding'
argument in the "USING ENCODING" clause of the CREATE DATABASE
statement.  The default is not to pass any "USING ENCODING" clause.
Valid encodings: 'utf-8' and 'iso-8859-1'.



.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 4 " Could not open file"
.TP
.BR 5 " Database error"
.TP
.BR 6 " Compiler error (error in MQL input)"
.SH AVAILABILITY
Available from http://emdros.org/.
.SH AUTHORS
Written Ulrik Sandborg-Petersen (ulrikp@emdros.org).


NEWS
This file is for release-notes for Emdros.

- *** Version 3.7.0 *** 

Release notes for Emdros 3.7.0
==============================

- Emdros has been relicensed under the MIT license.

- All copyright in Emdros and related documentation and build scripts
  has been assigned to Ulrik Sandborg-Petersen's holding company,
  Sandborg-Petersen Holding ApS.  This is so as to be able to serve
  customers better with support and other Emdros-related help.


- *** Version 3.6.1 *** 

Release notes for Emdros 3.6.1
==============================

Emdros version 3.6.1 is a maintenance release with the following
summary of changes since 3.6.0:

- The MQL query language has been extended with the ability to compare
  features with type LIST OF INTEGER, LIST OF ID_D, and LIST OF
  <enum-type> with an explicit list (including the empty list) using
  the "=" comparison operator.

  For example:

  SELECT ALL OBJECTS
  WHERE
  [Clause_atom
     daughters = () // Test for empty list
     OR
     daughters = (12345,67891) // Test for explicit, non-empty list
  ]
  GO

- Detection of PHP 5 and PHP 7 on macOS now works correctly.

- Compileability on Solaris 10 with Sun's CC C++ compiler has been
  restored.

- Compatibility with wxWidget 2.8.12 has been restored.


Enjoy!


Ulrik Sandborg-Petersen



- *** Version 3.6.0 *** 

Release notes for Emdros 3.6.0
==============================

The release notes below cover some of the important changes in Emdros
since 3.5.0.

- SQLite2 support has been removed, as version 2 of SQLite has not
  been updated by the SQLite authors for many, many years.  Also,
  SQLite2 support in Emdros has been been deprecated for some time.

- The MQL subsystem has a better query planner, which is able to
  choose the "Innermost First" strategy, rather than the default
  "Outermost First" strategy, under many more circumstances.

- The Emdros MySQL backend now has support for transactions, if
  available in the MySQL server.

- Assorted performance improvements to the following subsystems:

  - XML parsing (renderobjects, for example)
  - MQL

- Node.JS support has been removed.  The reason is that the latest
  SWIG version (3.0.12) does not support the latest Node.  Also, the
  Node.JS bindings never really worked.

- PCRE has been upgraded.

- SQLite3 has been upgraded.

- JSON output now escapes string characters correctly.

- Compiling Linux .rpm packages works on recent systems (CentOS 7,
  Fedora 28).

- Compiling Linux .deb packages works on recent systems (Ubuntu 16.04,
  Ubuntu 18.04, Debian 9.4).

- Building on Windows supports the latest Visual Studio.

- Lots of compiler warnings have been removed.

- the harvest template language has new constructs:

  - {{ inccounter <string> var <string> }}
  - {{ deccounter <string> var <string> }}

- The Full Text Search FTS3 subsystem is now able to obey
  largest_proximity, and to use * and ? in the query terms.

- A new program was added, bptdetach(1), which enables detaching
  "payloads" from BPT databases.

- The BPT backend (if available -- ask Ulrik for the source code
  licensing terms if you are interested) now supports LZMA compression
  of payloads.

- Source code using wxWidgets has been updated to support wxWidgets
  2.8, 3.0, and 3.1.

- The Emdros PostgreSQL backend now adds "TEMPLATE template0" when
  doing CREATE DATABASE.  This is so as to avoid using template1,
  which may conflict with the encoding used (and does so on Debian
  9.4).

- The amalgamation output has been split into several files.  This is
  because the Android NDK MIPS target cannot compile the old,
  monolithic amalgamation files.



- *** Version 3.5.0 *** 

Release notes for Emdros 3.5.0
==============================

The focus-areas of this release include:

- Enhancements to the core MQL
- Bugfixes to the core MQL.
- Enhancements to the render library
- Enhancements to the Full Text Search subsystem
- Enhancements to SWIG bindings
- Enhancements to the build system

Please consult the ChangeLog (going back to version 3.4.1.pre01) for
the specifics.  You can also consult the MQL Programmer's Guide for
the enhancements to the core MQL, and the Programmer's Reference Guide
for the enhancements to the Render library and the FTS subsystem.

Briefly, the core MQL has been enhanced as follows:

- SELECT ALL OBJECTS has a new WITH MAX RANGE clause, helping to limit
  otherwise runaway queries such as [Word]..[Word].

- SELECT ALL OBJECTS now sports an [UnorderedGroup] block.

- GET AGGREGATE FEAUTURES is a new statement, similar to SQL's MIN(),
  MAX(), SUM(), and COUNT() aggregate functions.


- *** Version 3.4.0 *** 

Release notes for Emdros 3.4.0
==============================

The focus-areas of this release include:

- Support for iOS and Android
- A smoother user experience for the GUI applications
- Enhancements to the MQL query language
- Documentation improvements
- Build improvements on all platforms
- Speed improvements


Binaries for Windows(R) and Mac OS X(R)
---------------------------------------

As usual, binaries are available for Mac OS X and Windows(R).

The Windows binaries are available in both x86 and x64 versions.  They
have built-in support for MySQL, SQLite 2, and SQLite 3.  The x86
binaries are built with Visual Studio Express 2010.  The x64 binaries
are built with Windows SDK 6.0A.

The Mac OS X binaries are Universal binaries (i386 and x86_64) running
on Mac OS X 10.6 (Snow Leopard), 10.7 (Lion), and 10.8 (Mountain
Lion), 10.9 (Mavericks), and possibly later versions of Mac OS X.
They do not have support for either MySQL or PostgreSQL; Only SQLite 2
and SQLite 3 are supported in the Mac OS X binaries.  You might be
able to compile the sources with support for MySQL yourself, and
possibly also PostgreSQL.

Linux binaries

src/mqldump.cpp
/*
 * mqldump.cpp
 *
 * A tool to dump Emdros databases in MQL
 * Created: 5/1-2001 (1st of May, 2001)
 * Last update: 5/11-2018
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 4 : Could not open file
 * 5 : Database error
 * 6 : Compiler error
 *
 */

/*
 * Copyright (C) 2001-2018     Ulrik Sandborg-Petersen
 * Copyright (C) 2018-present  Sandborg-Petersen Holding ApS, Denmark
 *
 * Licensed under the MIT License.
 *
 * Please see the file COPYING in the root of the sources for more details.
 *
 */




#include <emdros-config.h>
// Do this because emdros-config might #define malloc,
// in which case some systems will fail to use the real 
// malloc, thus causing build errors.
#undef malloc

#include <emdros-lconfig.h>
#include <encryption.h>
#include <iostream>
#include <string>
#include <environment_emdros.h>
#include <mql_exporter.h>
#include <opt.h>
#include <messages.h>



void show_version(std::ostream& ostr)
{
	ostr << "mqldump from Emdros version " << EMDROS_VERSION << std::endl;
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "   mqldump [options] database [object type names...] " << std::endl;
	ostr << "   mqldump (--help|-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	ostr << "   -v , --verbose       Be verbose" << std::endl;
	printUsageStandardArguments(ostr);
	ostr << "   -o filename          Dump to file rather than stdout" << std::endl;
	ostr << "   -n , --no-create-db            Don't issue statements to create and use db" << std::endl;
	ostr << "   --encoding encoding            Give db encoding in CREATE DATABASE stmnt." << std::endl
	     << "                                  (no effect if -n is used)" << std::endl;
	ostr << "   -i , --no-create-info          Don't issue statements to create object" << std::endl;
	ostr << "                                  types and enums (i.e., meta-data)." << std::endl;
	ostr << "   -t , --no-create-object-types  Don't issue statements to create object" << std::endl
	     << "                                  types" << std::endl;
	ostr << "   -e , --no-create-enums         Don't issue statements to create enums" << std::endl;
	ostr << "   -m , --no-create-monad-sets    Don't issue statements to create arbitrary" << std::endl;
	ostr << "                                  monad sets." << std::endl;
	ostr << "   -d , --no-create-objects       Don't issue statements to create" << std::endl 
	     << "                                  object data" << std::endl;
	ostr << "   --no-create-transactions       Don't issue transaction statements." << std::endl;
	ostr << "   --batch-create-objects         Issue CREATE OBJECTS WITH OBJECT TYPE" << std::endl
	     << "                                  statements rather than individual CREATE" << std::endl
	     << "                                  OBJECTS statements." << std::endl;
	ostr << "   --batch-size number            Set number of objects in each CREATE OBJECTS" << std::endl
	     << "                                  WITH OBJECT TYPE statement. Implies" << std::endl
	     << "                                  --batch-create-objects." << std::endl
	     << "   --start number                 Start dumping from the specified monad." << std::endl
	     << "   --end number                   Finish dumping at the specified monad." << std::endl;
	ostr << "DEFAULTS:  ";
	printUsageDefaultsOfStandardArguments(ostr);
	ostr << "--start <MIN_M>  --end <MAX_M>" << std::endl
	     << "--batch-size " << DEFAULT_BATCH_SIZE 
	     << " (but batch-creation doesn't happen by default.)"<< std::endl;
}

int main(int argc, char* argv[])
{
	// Set defaults
	eOutputKind output_kind = kOKConsole;
	MQLExportEnv de;
	eBackendKind backend_kind;
	eCharsets dummy_charset = kCSISO_8859_1;

	addStandardArguments(false // Do not add -e | --encoding
			     );

	addOption("-v", "--verbose");

	addOption("--no-create-transactions", "--no-create-transactions");

	addOption("-n", "--no-create-db");

	addOption("-i", "--no-create-info");

	addOption("--batch-create-objects", "--batch-create-objects");

	addOption("-e", "--no-create-enums");

	addOption("-m", "--no-create-monad-sets");

	addOption("-t", "--no-create-object-types");

	addOption("-d", "--no-create-objects");

	addOption("--encoding", "--encoding", true, 
		  "", // Default value
		  "ERROR: --encoding must have an argument.\n"
		  "       example: --encoding 'utf-8' or --encoding 'iso-8859-1'\n");

	addOption("-o", "--output", true, 
		  "", // Default value
		  "ERROR: -o and --output must have a filename as their argument.\n"
		  "       example: -o dump.mql\n");

	addOption("--start", "--start", true, 
		  "", // Default value
		  "ERROR: --start must have a monad as its argument\n"
		  "       example: --start 1\n");

	addOption("--end", "--end", true, 
		  "", // Default value
		  "ERROR: --end must have a monad as its argument\n"
		  "       example: --end 10000\n");

	addOption("--batch-size", "--batch-size", true, 
		  long2string(DEFAULT_BATCH_SIZE).c_str(), // Default value
		  "ERROR: --batch-size must have a number as its argument\n"
		  "       example: --end 50000\n");


	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  de.hostname,
					  de.user,
					  de.password,
					  backend_kind,
					  dummy_charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}

		if (getArgumentPresent("-v")) {
			de.bVerbose = true;
		}

		if (getArgumentPresent("--no-create-transactions")) {
			de.bUseTransactions = false;
		}

		if (getArgumentPresent("-n")) {
			de.bCreateAndUseDB = false;
		}

		if (getArgumentPresent("-i")) {
			de.bCreateEnums = false;
			de.bCreateObjectTypes = false;
		}

		if (getArgumentPresent("--batch-create-objects")) {
			de.bBatchCreateObjects = true;
		}

		if (getArgumentPresent("-e")) {
			de.bCreateEnums = false;
		}

		if (getArgumentPresent("-m")) {
			de.bCreateArbitraryMonadSets = false;
		}

		if (getArgumentPresent("-t")) {
			de.bCreateObjectTypes = false;
		}

		if (getArgumentPresent("-d")) {
			de.bCreateObjectData = false;
		}

		// -o | --output
		getArgumentValue("-o", de.filename);

		if (getArgumentPresent("--encoding")) {
			de.bDumpEncoding = true;
			getArgumentValue("--encoding", de.encoding);
			eCharsets charset;
			bool bShowError = !string2charset(de.encoding, charset);
			if (!bShowError) {
				if (charset != kCSUTF8
				    && charset != kCSISO_8859_1) {
					bShowError = true;
				}
			}
			if (bShowError) {
				std::cerr << "ERROR: The argument of --encoding must be a valid encoding.\n"
					  << "       Valid encodings:\n"
					  << "          - utf-8\n"
					  << "          - iso-8859-1" << std::endl;
				return 1;
			}
		}


		// --start
		std::string start_string;
		if (getArgumentValue("--start", start_string)) {
			if (string_is_number(start_string)) {
				de.m_start = string2long(start_string);
			} else {
				print_usage(std::cerr);
				std::cerr << "ERROR: The argument of --start must be a number, and not '" << start_string << "'." << std::endl;
				return 1;
			}
		}


		// --end
		std::string end_string;
		if (getArgumentValue("--end", end_string)) {
			if (string_is_number(end_string)) {
				de.m_end = string2long(end_string);
			} else {
				print_usage(std::cerr);
				std::cerr << "ERROR: The argument of --end must be a number, and not '" << end_string << "'." << std::endl;
				return 1;
			}
		}

		// --batch-size
		std::string batch_size_string;
		if (getArgumentValue("--batch-size", batch_size_string)) {
			if (string_is_number(batch_size_string)) {
				de.nBatchSize = string2long(batch_size_string);
				de.bBatchCreateObjects = true;
			} else {
				print_usage(std::cerr);
				std::cerr << "ERROR: The argument of --batch-size must be a number, and not '" << end_string << "'." << std::endl;
				return 1;
			}
		}


		// There must be surplus arguments
		if (surplus_arguments.size() == 0) {
			print_usage(std::cerr);
			std::cerr << "ERROR: You must at least specify a database on the command line.\n"
				  << "       If you wish, you can also specify a number of object types to dump,\n"
				  << "       after the database." << std::endl;
			return 1;
		} else if (surplus_arguments.size() == 1) {
			// If there is only one surplus argument, it is the DB name
			de.db = *surplus_arguments.begin();
		} else {
			// Otherwise, the first is the DB name, and the rest are object types.
			de.db = *surplus_arguments.begin();
			surplus_arguments.erase(surplus_arguments.begin());
			de.object_types = surplus_arguments;
		}
	}

	
	// Make connection
	EmdrosEnv *pEE = new EmdrosEnv(output_kind, kCSISO_8859_1,
				       de.hostname,
				       de.user, de.password,
				       de.db,
				       backend_kind);
	// Zero-fill password
	zeroFillString(de.password);

	// Check that we connected
	if (!pEE->connectionOk()) {
		emdrosMessageConnectionNotOK(&std::cerr, backend_kind);
		delete pEE;
		return 2;
	}

	int nResult = 3;
	std::string message;
	try {
		if (de.filename == "") {
			MQLExporter exporter(de, pEE);
			nResult = exporter.doIt(message);
		} else {
			MQLExporter exporter(de, pEE, de.filename);
			nResult = exporter.doIt(message);
		}
		if (nResult != 0) {
			std::cerr << pEE->getDBError() << std::endl; 
			std::cerr << pEE->getCompilerError() << std::endl;
			std::cerr << message << std::endl;
		}
	} catch (EMdFDBException& e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	} 

	// Clean up
	delete pEE;
	return nResult;
}


